ledger/internal/evalprefetcher.go

```diff
@@ -0,0 +1,520 @@
+	balancesCount int
+	// resources contains the loaded resources each of the transaction groups have
+	resources []loadedResourcesEntry
+	// resourcesCount is the number of resources that nees to be loaded per transaction group
+	resourcesCount int
+	// incompleteCount is the number of resources+balances still pending and need to be loaded
+	incompleteCount int64
+	// err indicates the error code ( if any ) that was received from the ledger during execution
+	err error
+}
+
+// preloaderTask manage the loading of a single element, whether it's a resource or an account address.
+type preloaderTask struct {
+	// account address to fetch
+	address *basics.Address
+	// resource id
+	creatableIndex basics.CreatableIndex
+	// resource type
+	creatableType basics.CreatableType
+	// a list of transaction group tasks that depends on this address
+	groups []*groupTask
+	// a list of indices into the groupTask.balances or groupTask.resources where the address would be stored
+	groupIndices []int
+}
+
+// preloaderTaskQueue is a dynamic linked list of enqueued entries, optimized for non-syncronized insertion and
+// syncronized extraction
+type preloaderTaskQueue struct {
+	next               *preloaderTaskQueue
+	used               int
+	entries            []*preloaderTask
+	baseIdx            int
+	maxTxnGroupEntries int
+}
+
+func allocPreloaderQueue(count int, maxTxnGroupEntries int) preloaderTaskQueue {
+	return preloaderTaskQueue{
+		entries:            make([]*preloaderTask, count*2+maxTxnGroupEntries*2),
+		maxTxnGroupEntries: maxTxnGroupEntries,
+	}
+}
+
+// enqueue places the queued entry on the queue, returning the latest queue
+// ( in case the current "page" ran out of space )
+func (pq *preloaderTaskQueue) enqueue(t *preloaderTask) {
+	pq.entries[pq.used] = t
+	pq.used++
+	return
+}
+
+func (pq *preloaderTaskQueue) expand() *preloaderTaskQueue {
+	if cap(pq.entries)-pq.used < pq.maxTxnGroupEntries {
+		pq.next = &preloaderTaskQueue{
+			entries: make([]*preloaderTask, cap(pq.entries)*2),
+			used:    0,
+			baseIdx: pq.baseIdx + pq.used,
+		}
+		return pq.next
+	}
+	return pq
+}
+
+func (pq *preloaderTaskQueue) getTaskAtIndex(idx int) (*preloaderTaskQueue, *preloaderTask) {
+	localIdx := idx - pq.baseIdx
+	if pq.used > localIdx {
+		return pq, pq.entries[localIdx]
+	}
+	if pq.next != nil {
+		return pq.next.getTaskAtIndex(idx)
+	}
+	return pq, nil
+}
+
+type accountCreatableKey struct {
+	address basics.Address
+	cidx    basics.CreatableIndex
+}
+
+func loadAccountsAddAccountTask(addr *basics.Address, wg *groupTask, accountTasks map[basics.Address]*preloaderTask, queue *preloaderTaskQueue) {
+	if addr.IsZero() {
+		return
+	}
+	if task, have := accountTasks[*addr]; !have {
+		task := &preloaderTask{
+			address:      addr,
+			groups:       make([]*groupTask, 1, 4),
+			groupIndices: make([]int, 1, 4),
+		}
+		task.groups[0] = wg
+		task.groupIndices[0] = wg.balancesCount
```

Comment:
not sure I understand how `balancesCount` works. it is 0 on the beginning for all groups except the first one. Why would we set 0 in a new task, and not 1 ?

Reply:
There are two types of objects : `preloaderTask` and `groupTask`. 
A single `preloaderTask` would load a single resource that can be used by many `groupTask`.
The balancesCount on each `groupTask` keeps track of the index where the `preloaderTask` would write the resulting resource onto.
This mechanism allows us to avoid taking any locks by planning upfront where in the memory each output would go.

Reply:
I see, it is used as an index and then as a ops count. thanks